[WPE][Skia] Build Skia on WPE-Android

zhani opened the pull request:
#### 9bdd310f0173111eb6060a447aa623d62816cb7c
<pre>
[WPE][Skia] Build Skia on WPE-Android
<a href="https://bugs.webkit.org/show_bug.cgi?id=281452">https://bugs.webkit.org/show_bug.cgi?id=281452</a>

Reviewed by Adrian Perez de Castro.

Build and enable Skia on WPE-Android

* Source/ThirdParty/skia/CMakeLists.txt:
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
(WebCore::FontCache::fontManager const):

Canonical link: <a href="https://commits.webkit.org/285349@main">https://commits.webkit.org/285349@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e6c35c4646b37b82cd6c2aa51e9143075ea43f0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72347 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51768 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25135 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23556 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74462 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59572 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23378 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/76524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15506 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75414 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46879 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62303 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43532 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19775 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21906 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20132 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78194 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16588 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19270 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16635 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62321 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15979 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12974 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/6608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47566 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/48635 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/49923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48378 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
```diff
@@ -29,7 +29,11 @@
 #include "Font.h"
 #include "FontDescription.h"
 #include "StyleFontSizeFunctions.h"
+#if defined(ANDROID)
```
The [official answer](https://groups.google.com/g/android-ndk/c/cf9_f1SLXls/m/eI3UekPW4zEJ) seems to be that one should prefer `__ANDROID__` because it's always defined by the compiler, but in some places people have been using `ANDROID` as well... which explains why there's checks for both macros e.g. in ANGLE.

I would check both as well just in case:

```suggestion
#if defined(__ANDROID__) || defined(ANDROID)
```

zhani replied:
Fixed

aperezdc commented on Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
```diff
@@ -47,8 +51,13 @@ void FontCache::platformInit()
 {
-    if (!m_fontManager)
+    if (!m_fontManager) {
+#if defined(ANDROID)
```
Same here:

```suggestion
#if defined(__ANDROID__) || defined(ANDROID)
```

zhani replied:
Fixed

aperezdc commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -895,6 +901,19 @@ add_library(Skia STATIC
+    )
+else ()
+    target_sources(Skia PRIVATE src/ports/SkDebug_stdio.cpp)
+    target_sources(Skia PRIVATE src/ports/SkFontMgr_fontconfig.cpp)
```
I think we need to add a `target_link_libraries(Skia PRIVATE Fontconfig::Fontconfig)` here, and remove it from the list of libraries right below after the `endif`... As far as I remember Android does not use Fontconfig, or at least it's not exposed to the NDK.

zhani replied:
For now we build it for WPE-Android because of other dependencies from gstreamer

zhani replied:
But patch is now update according to this suggestion

aperezdc reviewed (commented):
Thanks for sending this! I was expecting a bigger patch to get Skia working on Android, so this is quite nice. There are a couple of small things to change, and one CMake doubt I have, but otherwise the patch looks good to me.

aperezdc commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -1,10 +1,18 @@
+if (ANDROID)
+    set(is_android TRUE)
+endif ()
+
```
Remove this:

```suggestion
```

zhani replied:
I added this because similar style was used in ANGLE CMake files

zhani replied:
Removed now

aperezdc commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -1,10 +1,18 @@
 
+if (is_android)
+find_package(Expat REQUIRED)
+endif ()
```
Check `ANDROID` directly + indentation:

```suggestion
if (ANDROID)
    find_package(Expat REQUIRED)
endif ()
```

zhani replied:
Fixed

aperezdc commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -895,9 +901,25 @@ add_library(Skia STATIC
     modules/skcms/src/skcms_TransformSkx.cc
 )
 
+if (is_android)
```
Check `ANDROID` directly here as well:

```suggestion
if (ANDROID)
```

zhani replied:
Fixed

aperezdc reviewed (approved):
Thanks for the update, could you apply the suggested changes to the CMake file? With that I think this is ready to land.

aperezdc commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -5,6 +5,10 @@ find_package(Freetype 2.9.0 REQUIRED)
 find_package(WebP REQUIRED COMPONENTS mux)
 
+if (ANDROID)
+    find_package(Expat REQUIRED)
```
Apparently this should be `EXPAT`, see [the documentation for the find-module](https://cmake.org/cmake/help/latest/module/FindEXPAT.html), and we do not have a `Source/cmake/FindExpat.cmake` file.

```suggestion
    find_package(EXPAT REQUIRED)
```

zhani replied:
Fixed and tested

aperezdc commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -895,9 +897,25 @@ add_library(Skia STATIC
+    target_sources(Skia PRIVATE src/ports/SkFontMgr_android_parser.cpp)
+
+    target_link_libraries(Skia PRIVATE
+        expat::expat
```
Here as well:

```suggestion
        EXPAT::EXPAT
```

zhani replied:
Fixed and tested

aperezdc reviewed (commented):
One last nit, sorry about the back-and-forth...

aperezdc reviewed (commented):
@zhani Thanks! This is now ready for landing, let's send it to the commit queue 👍🏼

zhani committed:
[WPE][Skia] Build Skia on WPE-Android
https://bugs.webkit.org/show_bug.cgi?id=281452

Reviewed by Adrian Perez de Castro.

Build and enable Skia on WPE-Android

* Source/ThirdParty/skia/CMakeLists.txt:
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
(WebCore::FontCache::fontManager const):

Canonical link: https://commits.webkit.org/285349@main
